### PR TITLE
add option to attach node using cluster uuid

### PR DESCRIPTION
**Pull request by OmkarDeshpande7**
## ISSUE(S):
[PMK-4315](https://platform9.atlassian.net/browse/PMK-4315)

## SUMMARY
- Added option to attach a node to a cluster using cluster uuid
- Would be helpful for CAPI automation using pf9ctl

## ISSUE TYPE
- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that may cause existing functionality to not work as expected)
- [x] This change requires a documentation update

## IMPACTED FEATURES/COMPONENTS:
attach-node

#### Autom
Screenshot 2022-03-15 at 3 24 47 PM
---
Screenshot 2022-03-15 at 3 24 59 PM
---
Screenshot 2022-03-15 at 3 25 33 PM
---
Screenshot 2022-03-15 at 3 29 50 PM
---
Screenshot 2022-03-15 at 3 36 12 PM

#### Reviewers
/cc @devidask27 @nikhilbandi @anupbarve @erwinvaneyk

### CODE REVIEW

File: pkg/qbert/qbert.go [hunk 1]
```diff
@@ -387,6 +388,42 @@ func (c QbertImpl) CheckClusterExists(name, projectID, token string) (bool, stri
...
+	}
+
+	req.Header.Set("X-Auth-Token", token)
+	req.Header.Set("Content-Type", "application/json")
+	resp, err := client.Do(req)
+	if err != nil {
+		return false, "", err
+	}
+	if resp.StatusCode != 200 {
+		return false, "", fmt.Errorf("Couldn't query the qbert Endpoint: %d", resp.StatusCode)
```
Comment (erwinvaneyk) [inline pkg/qbert/qbert.go:407]: * Errors should not be capitalized
* Avoid colloquialisms
```suggestion
		return false, "", fmt.Errorf("could not query the qbert Endpoint: %d", resp.StatusCode)
```
(same goes for other errors)

File: cmd/attachNode.go [hunk 2]
```diff
@@ -95,8 +99,20 @@ func attachNodeRun(cmd *cobra.Command, args []string) {
...
+	var cluster_uuid string
+	if clusterUuid != "" {
+		var exists bool
+		if exists, clusterName, err = c.Qbert.CheckClusterExistsWithUuid(clusterUuid, projectId, token); err != nil {
+			zap.S().Fatalf("unable to verify cluster using uuid", err.Error())
+		} else if exists {
+			cluster_uuid = clusterUuid
+		} else {
+			zap.S().Fatalf("cluster with given uuid does not exist")
+		}
```
Comment (erwinvaneyk) [inline cmd/attachNode.go:111]: I don't think a cluster can have an empty name, so this `exist` is a bit unnecessary.

File: cmd/attachNode.go [hunk 3]
```diff
@@ -46,8 +47,11 @@ var (
 func init() {
 	attachNodeCmd.Flags().StringSliceVarP(&masterIPs, "master-ip", "m", []string{}, "master node ip address")
 	attachNodeCmd.Flags().StringSliceVarP(&workerIPs, "worker-ip", "w", []string{}, "worker node ip address")
+	attachNodeCmd.Flags().StringVarP(&clusterUuid, "uuid", "u", "", "cluster uuid")
```
Comment (erwinvaneyk) [inline cmd/attachNode.go:50]: Try to make this a full sentence 😉

File: cmd/attachNode.go [hunk 4]
```diff
@@ -29,6 +29,7 @@ var (
 		Short: "Attaches node to kubernetes cluster",
 		Long:  "Attach nodes to existing cluster. At a time, multiple workers but only one master can be attached",
 		Args: func(attachNodeCmd *cobra.Command, args []string) error {
+			// even if '-u' option is specified use some dummy cluster name
```
Comment (erwinvaneyk) [inline cmd/attachNode.go:32]: I think the flag is already set as this point, so you can use it in this check and avoid this dummy thing

File: cmd/attachNode.go [hunk 5]
```diff
@@ -95,8 +99,20 @@ func attachNodeRun(cmd *cobra.Command, args []string) {
...
+		var exists bool
+		if exists, clusterName, err = c.Qbert.CheckClusterExistsWithUuid(clusterUuid, projectId, token); err != nil {
+			zap.S().Fatalf("unable to verify cluster using uuid", err.Error())
+		} else if exists {
+			cluster_uuid = clusterUuid
+		} else {
+			zap.S().Fatalf("cluster with given uuid does not exist")
+		}
+	} else {
+		_, cluster_uuid, _ = c.Qbert.CheckClusterExists(clusterName, projectId, token)
```
[the remark this answers is not in the document]
Comment (OmkarDeshpande7) [reply cmd/attachNode.go:113]: Hi @erwinvaneyk, good point. But I did not want to disturb the next bit of code where the variable name is already used in that fashion.

**Review by erwinvaneyk — APPROVED** [review]
Looks good, but a review from @devidask27 or @nikhilbandi is needed too

**Review by anupbarve — CHANGES_REQUESTED** [review]
- Please add screenshots of the test cases executed manually.
- README needs to be updated, since we are changing the CLI options.
- This would also need changes to the pf9ctl external facing docs. This needs to be tracked.
cc: @devidask27 @nikhilbandi

File: cmd/attachNode.go [hunk 6]
```diff
@@ -46,8 +49,11 @@ var (
 func init() {
 	attachNodeCmd.Flags().StringSliceVarP(&masterIPs, "master-ip", "m", []string{}, "master node ip address")
 	attachNodeCmd.Flags().StringSliceVarP(&workerIPs, "worker-ip", "w", []string{}, "worker node ip address")
+	attachNodeCmd.Flags().StringVarP(&clusterUuid, "uuid", "u", "", "uuid of a cluser to attach the node to")
```
Comment (anmolsachan) [inline cmd/attachNode.go:52]:
```suggestion
	attachNodeCmd.Flags().StringVarP(&clusterUuid, "uuid", "u", "", "uuid of the cluster to attach the node to.")
```

Comment (OmkarDeshpande7) [reply cmd/attachNode.go:52]: Done

File: cmd/attachNode.go [hunk 7]
```diff
@@ -46,8 +49,11 @@ var (
...
 	attachNodeCmd.Flags().StringVar(&attachconfig.MFA, "mfa", "", "MFA token")
 	rootCmd.AddCommand(attachNodeCmd)
+	// '-u' option for inetrnal use only for now
+	attachNodeCmd.Flags().MarkHidden("uuid")
```
Comment (anupbarve) [inline cmd/attachNode.go:56]: why is this hidden ?

File: pkg/qbert/qbert.go [hunk 8]
```diff
@@ -387,6 +388,43 @@ func (c QbertImpl) CheckClusterExists(name, projectID, token string) (bool, stri
...
+	req.Header.Set("X-Auth-Token", token)
+	req.Header.Set("Content-Type", "application/json")
+	resp, err := client.Do(req)
+	if err != nil {
+		return "", err
+	}
+	if resp.StatusCode == 400 {
+		return "", fmt.Errorf("cluster with given uuid does not exist: %d", resp.StatusCode)
+	} else if resp.StatusCode != 200 {
+		return "", fmt.Errorf("could not query the qbert Endpoint: %d", resp.StatusCode)
```
Comment (anupbarve) [inline pkg/qbert/qbert.go:409]: nit: 
%s/Endpoint/endpoint

**Review by anupbarve — CHANGES_REQUESTED** [review]

File: pkg/qbert/qbert.go (hunk 8 above)
Comment (OmkarDeshpande7) [reply pkg/qbert/qbert.go:409]: done

File: cmd/attachNode.go (hunk 7 above)
Comment (OmkarDeshpande7) [reply cmd/attachNode.go:59]: Hi @anupbarve, Thanks!
Initially, it was needed for internal usage and testing. There is no specific reason to mark this hidden. We can make the flag public, should we?

Comment (anupbarve) [reply cmd/attachNode.go:59]: I would like @anmolsachan to comment on this.
IMO, I see this as a value add that can be made public.

**Comment by OmkarDeshpande7** [comment]
Screenshot 2022-02-24 at 2 00 01 PM

**Comment by anmolsachan** [comment]
Hey @OmkarDeshpande7, Can you please elaborate on the need for this PR? Any engineering/customer-facing requirements?

**Comment by OmkarDeshpande7** [comment]
Hi @anmolsachan,
There aren't any customer-facing requirements that I am aware of. However, previously `pf9ctl` allowed only cluster name to be supplied as a parameter to `attach-node`. Now, keeping in mind that we allow clusters with the same names to be created in qbert, this causes `pf9ctl` to attach the node to the wrong cluster if two clusters of the same name are present (OR even fail).
This issue was noticed during testing of CAPI related work (Since we are using pf9ctl based automation to onboard a node to DU). So, having UUID as a parameter would help to reduce unintentional errors.
Just highlighting again, there is no specific reason to mark this flag as hidden.

**Comment by anmolsachan** [comment]
Thanks, @OmkarDeshpande7, I think that is a good enough reason. I would let @anupbarve @devidask27 provide final approval.

File: pkg/qbert/qbert.go [hunk 9]
```diff
@@ -29,7 +29,9 @@ type Qbert interface {
 	AuthoriseNode(nodeUuid, token string) error
 	GetNodePoolID(projectID, token string) (string, error)
 	CheckClusterExists(Name, projectID, token string) (bool, string, error)
-	CheckClusterExistsWithUuid(uuid, projectID, token string) (string, error)
```
Comment (devidask27) [inline pkg/qbert/qbert.go]: Why this is removed ?

Comment (OmkarDeshpande7) [reply pkg/qbert/qbert.go]: It has been restored

**Comment by devidask27** [comment]
@OmkarDeshpande7 Also update the README.

File: pkg/qbert/qbert.go [hunk 10]
```diff
@@ -398,6 +399,43 @@ func (c QbertImpl) CheckClusterExists(name, projectID, token string) (bool, stri
...
+	if err != nil {
+		return "", err
+	}
+
+	if payload["uuid"] == uuid {
+		cluster_name := payload["name"].(string)
+		return cluster_name, nil
+	}
+
+	return fmt.Sprintf("error finding cluster with uuid %s", uuid), nil
```
Comment (devidask27) [inline pkg/qbert/qbert.go:436]: err should not be nil in this case, return some error message.

**Review by devidask27 — CHANGES_REQUESTED** [review]

File: pkg/qbert/qbert.go (hunk 10 above)
Comment (OmkarDeshpande7) [reply pkg/qbert/qbert.go]: Done

**Review by devidask27 — APPROVED** [review]
LGTM

File: cmd/attachNode.go [hunk 11]
```diff
@@ -95,8 +101,20 @@ func attachNodeRun(cmd *cobra.Command, args []string) {
 	}
 	projectId := auth.ProjectID
 	token := auth.Token
+	var clusterUUID string
+	if clusterUuid != "" {
```
[the remark this answers is not in the document]
Comment (OmkarDeshpande7) [reply cmd/attachNode.go:104]: Done

**Review by anupbarve — APPROVED** [review]
lgtm

**Pull request merged**